### PR TITLE
Fix `SymbolicHamiltonian.expectation` for different nqubits

### LINKED ISSUE #935: Wrong results using `SymbolicHamiltonian`
I have recently found a bug in the `SymbolicHamiltonian` class, whereby expectation values of certain observables are not calculated correctly when using the `qibojit` backend (version `0.0.10`).

In particular, here is an example where this behavior shows up. It is a 2-qubit circuit where I apply an `RXX(theta)` gate and then measure the `Z(0)` observable. The expected value is easy to compute, it should be `cos(theta)`. However, `qibo` (version `0.1.14`) returns a constant value of 1 for all values of `theta`.

The code to reproduce this issue is:

```python
# CREATE CIRCUIT
circ = Circuit(2)
circ.add(gates.RXX(0, 1, theta=0))

# DEFINE SYMBOLIC HAMILTONIAN
operator = SymbolicHamiltonian(Z(0))

# COLLECT RESULTS
results = []
for t in np.linspace(0, 2*np.pi, 1000):
    circ.set_parameters([t])
    state = circ().state()
    results.append(operator.expectation(state))
    
# PLOT RESULTS
import matplotlib.pyplot as plt
plt.plot(np.linspace(0, 2*np.pi, 1000), results)
plt.plot(np.linspace(0, 2*np.pi, 1000), np.cos(np.linspace(0, 2*np.pi, 1000)))
plt.show()
```

If we now change `operator = SymbolicHamiltonian(Z(0))` by `operator = SymbolicHamiltonian(Z(0)*I(1))`, the code works as expected. Moreover, when using the `numpy` backend the snippet above doesn't work unless the `I(1)` is included.

This shows that the problem is in the `Z(0)`, not in the circuit. The code also works as expected if we decompose the `RXX` gate in terms of `H`, `RZ` and `CNOT` gates.

Any clue about what is going on? I think this is an important issue, as people may be finding wrong results in their research projects.

Also, this issue went through the tests, so I guess those should be updated as well.

### PR DESCRIPTION
**Pull request by stavros11**
Fixes #935. Specifically:
* `nqubits` attribute was added to `SymbolicHamiltonian` as suggested in https://github.com/qiboteam/qibo/issues/935#issuecomment-1592312035. Note that the effect of this is still equivalent to padding, for example `SymbolicHamiltonian(Z(0) * Z(1), nqubits=4)` is equivalent to `SymbolicHamiltonian(Z(0) * Z(1) * I(2) * I(3))`.
* A `ValueError` is raised if there is a mismatch between the number of qubits in the state and the Hamiltonian in expectation value calculations.
* A documentation example about `expectation_from_samples` was added (@AlejandroSopena let me know if that is okay).

Checklist:
- [ ] Reviewers confirm new code works as expected.
- [x] Tests are passing.
- [x] Coverage does not decrease.
- [x] Documentation is updated.

### CODE REVIEW

File: doc/source/code-examples/advancedexamples.rst [hunk 1]
```diff
@@ -1897,3 +1897,74 @@ constructing each symbol:
...
+
+    circuit = Circuit(4)
+    circuit.add(gates.H(i) for i in range(4))
+    c.add(gates.CNOT(0, 1))
+    c.add(gates.CNOT(1, 2))
+    c.add(gates.CNOT(2, 3))
+
+    hamiltonian = XXZ(4)
+
+    result = circuit()
```
Comment (igres26) [inline doc/source/code-examples/advancedexamples.rst:1956]:
```suggestion
    result = circuit(nshots=1024)
```
I think that for the expectations from samples we should make it clear that we are doing a certain number of shots, even if there is a default value.

Comment (stavros11) [reply doc/source/code-examples/advancedexamples.rst]: Thanks, I missed this. This code will throw an error anyway (`result.frequencies()` won't be available unless you specify `nshots`), so is strange why the doctest didn't fail...

**Review by igres26 — APPROVED** [review]

File: doc/source/code-examples/advancedexamples.rst [hunk 2]
```diff
@@ -1897,3 +1897,74 @@ constructing each symbol:
...
+    circuit = Circuit(4)
+    circuit.add(gates.H(i) for i in range(4))
+    c.add(gates.CNOT(0, 1))
+    c.add(gates.CNOT(1, 2))
+    c.add(gates.CNOT(2, 3))
+
+    hamiltonian = XXZ(4)
+
+    result = circuit(nshots=1024)
+    expectation_value = hamiltonian.expectation_from_samples(result.frequencies())
```
Comment (AlejandroSopena) [inline doc/source/code-examples/advancedexamples.rst:1957]:
```python
from qibo import gates
from qibo.hamiltonians import Z
from qibo.models import Circuit

circuit = Circuit(4)
circuit.add(gates.H(i) for i in range(4))
circuit.add(gates.CNOT(0, 1))
circuit.add(gates.CNOT(1, 2))
circuit.add(gates.CNOT(2, 3))
circuit.add(gates.M(*range(4)))

hamiltonian = Z(4)

result = circuit(nshots=1024)
expectation_value = hamiltonian.expectation_from_samples(result.frequencies())
```

**Review by AlejandroSopena — COMMENTED** [review]
Thanks @stavros11 for adding the example. The tests are not passing because `expectation_from_samples()` only works when the observable is diagonal in the z basis as you mentioned, and `XXZ()` is not. I have pending to generalize this function #852 and I will do it at some point in August.

File: doc/source/code-examples/advancedexamples.rst (hunk 2 above)
Comment (stavros11) [reply doc/source/code-examples/advancedexamples.rst]: Thanks @AlejandroSopena. Indeed I even wrote that only diagonal Hamiltonians are supported for now but didn't follow it in the example.

File: doc/source/code-examples/advancedexamples.rst [hunk 3]
```diff
@@ -1897,3 +1897,74 @@ constructing each symbol:
...
+
+    from qibo import gates
+    from qibo.models import Circuit
+    from qibo.hamiltonians import Z
+
+    circuit = Circuit(4)
+    circuit.add(gates.H(i) for i in range(4))
+    c.add(gates.CNOT(0, 1))
+    c.add(gates.CNOT(1, 2))
+    c.add(gates.CNOT(2, 3))
```
Comment (AlejandroSopena) [inline doc/source/code-examples/advancedexamples.rst:1952]:
```python
circuit.add(gates.CNOT(0, 1))
circuit.add(gates.CNOT(1, 2))
circuit.add(gates.CNOT(2, 3))
circuit.add(gates.M(*range(4)))
```

File: doc/source/code-examples/advancedexamples.rst [hunk 4]
```diff
@@ -1897,3 +1897,74 @@ constructing each symbol:
...
+
+    from qibo import gates
+    from qibo.models import Circuit
+    from qibo.hamiltonians import XXZ
+
+    circuit = Circuit(4)
+    circuit.add(gates.H(i) for i in range(4))
+    c.add(gates.CNOT(0, 1))
+    c.add(gates.CNOT(1, 2))
+    c.add(gates.CNOT(2, 3))
```
Comment (AlejandroSopena) [inline doc/source/code-examples/advancedexamples.rst:1922]:
```python
  circuit.add(gates.CNOT(0, 1))
  circuit.add(gates.CNOT(1, 2))
  circuit.add(gates.CNOT(2, 3))
```

**Review by AlejandroSopena — COMMENTED** [review]
Thanks. Just a few more comments. When adding the three CNOTs, you have to replace `c` --> `circuit` To use `expectation_from_samples()` we have to add measurement gates because the expected value is calculated from frequencies.

**Comment by stavros11** [comment]
Thanks for the fixes @AlejandroSopena. I had them locally in my computer but forgot to push. I hope it is okay now.

**Review by AlejandroSopena — APPROVED** [review]

**Pull request merged**